[content-visibility] Add support for ContentVisibilityAutoStateChangeEvent

rwlbuis opened the pull request:
#### 5b575ff841de2b2490b8707d303e068c48641e93
<pre>
[content-visibility] Add support for ContentVisibilityAutoStateChangeEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=258880">https://bugs.webkit.org/show_bug.cgi?id=258880</a>

Reviewed by Tim Nguyen.

Support ContentVisibilityAutoStateChangeEvent, see:
<a href="https://drafts.csswg.org/css-contain/#content-visibility-auto-state-change">https://drafts.csswg.org/css-contain/#content-visibility-auto-state-change</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-state-changed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-state-changed-first-observation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-state-changed-removed-expected.txt: Added.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.cpp: Added.
(WebCore::ContentVisibilityAutoStateChangeEvent::ContentVisibilityAutoStateChangeEvent):
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.h: Added.
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.idl: Added.
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateRelevancyOfContentVisibilityElements): bail out early if no update is specified
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isRelevantToUser const):
(WebCore::Element::contentRelevancy const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::contentRelevancy const):
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/html/HTMLAttributeNames.in:

Canonical link: <a href="https://commits.webkit.org/267340@main">https://commits.webkit.org/267340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f3e793502cb8310ec3a5f4e8b83ba09e5ea9713

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16306 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16625 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17047 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18074 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15293 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16495 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19714 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16744 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17673 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16500 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16940 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13990 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18841 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14766 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21577 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15172 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14930 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18455 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15520 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13158 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14738 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3906 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15347 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.idl:
```diff
@@ -0,0 +1,35 @@
+ */
+
+[
+    Exposed=Window
```
```suggestion
    EnabledBySetting=CSSContentVisibilityEnabled,
    Exposed=Window
```

rwlbuis replied:
Makes sense, done.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -123,10 +125,13 @@ DidUpdateAnyContentRelevancy ContentVisibilityDocumentState::updateRelevancyOfCo
                 setRelevancyValue(ContentRelevancy::IsInTopLayer, hasTopLayerinSubtree(*target));
 
-            if (oldRelevancy == newRelevancy)
+            if (oldRelevancy == newRelevancy && oldRelevancy)
```
Why is this correct? If both old & new are both empty then there is no event to be emitted right?

rwlbuis replied:
The spec is not too clear about it, but both gecko and chromium dispatch this event when first determining rendering state, independent whether that turns out skipped or not. To make sure we are catching the first rendering state case I made the content relevancy std::optional.

nt1m replied:
Can we add a WPT for it?

nt1m replied:
Oh I guess that's imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-state-changed-first-observation.html

rwlbuis replied:
Correct :)

rwlbuis committed:
[content-visibility] Add support for ContentVisibilityAutoStateChangeEvent
https://bugs.webkit.org/show_bug.cgi?id=258880

Reviewed by Tim Nguyen.

Support ContentVisibilityAutoStateChangeEvent, see:
https://drafts.csswg.org/css-contain/#content-visibility-auto-state-change

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-state-changed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-state-changed-first-observation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-state-changed-removed-expected.txt: Added.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.cpp: Added.
(WebCore::ContentVisibilityAutoStateChangeEvent::ContentVisibilityAutoStateChangeEvent):
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.h: Added.
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.idl: Added.
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateRelevancyOfContentVisibilityElements): bail out early if no update is specified
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isRelevantToUser const):
(WebCore::Element::contentRelevancy const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::contentRelevancy const):
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/html/HTMLAttributeNames.in:

Canonical link: https://commits.webkit.org/267340@main
